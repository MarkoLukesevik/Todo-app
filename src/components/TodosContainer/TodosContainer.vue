<script setup lang="ts">
import { ref } from 'vue';
import type { Ref } from 'vue';

import type { Todo } from '@/models/Todo';

import CreateTodo from '@/components/TodosContainer/components/CreateTodo.vue';
import TodoList from './components/TodoList/TodoList.vue';
import type { AppThemeEnum } from '@/models/AppThemeEnum';

defineProps<{
    appTheme: AppThemeEnum;
}>();

const todos: Ref<Todo[]> = ref([]);

const addNewTodo = (todo: Todo): void => {
    todos.value.push(todo);
}

const toggleTodo = (id: string): void => {
    const foundTodo = todos.value.find((todo: Todo) => todo.id === id);
    if (foundTodo)
      foundTodo.isComplete = !foundTodo.isComplete;
}

const deleteTodo = (id: string): void => {
    todos.value = todos.value.filter((todo: Todo) => todo.id !== id);
}
</script>

<template>
  <create-todo :app-theme="appTheme" @create-todo="addNewTodo" />
  <todo-list
      :app-theme="appTheme"
      :todos="todos"
      @toggle-todo="toggleTodo"
      @delete-todo="deleteTodo"
  />
</template>

<style scoped></style>
