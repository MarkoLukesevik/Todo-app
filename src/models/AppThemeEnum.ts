export enum AppThemeEnum {
    Light = 0,
    Dark = 1
}
