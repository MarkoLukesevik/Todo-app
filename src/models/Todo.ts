export interface Todo {
  id: string
  title: string
  isComplete: false
}
