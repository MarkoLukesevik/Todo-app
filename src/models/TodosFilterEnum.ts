export enum TodosFilterEnum {
    All = 0,
    Active = 1,
    Completed = 2
}