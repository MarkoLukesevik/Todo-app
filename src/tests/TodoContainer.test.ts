import { describe, it, beforeEach, expect } from "vitest";
import { shallowMount } from "@vue/test-utils";

import { mockLocalStorage } from "./mockLocalStorage/mockLocalStorage";
import { TodosFilterEnum } from "@/models/TodosFilterEnum";

import TodosContainer from "@/components/TodosContainer/TodosContainer.vue";
import { AppThemeEnum } from "@/models/AppThemeEnum";

Object.defineProperty(window, 'localStorage', { value: mockLocalStorage });

describe('Test suite for todo container component', () => {
    let wrapper: any;

    beforeEach(() => {
        localStorage.clear();
        wrapper = shallowMount(TodosContainer, {
            props: {appTheme: AppThemeEnum.Light}
        });
    });

    it('Should add a new todo and update local storage', async () => {
        const newTodo = { id: '1', text: 'New Todo', isComplete: false };

        await wrapper.vm.addNewTodo(newTodo);

        expect(wrapper.vm.todos).toContainEqual(newTodo);
        expect(wrapper.vm.currentTodos).toContainEqual(newTodo);

        const storedTodos = JSON.parse(localStorage.getItem('todos') || '[]');
        expect(storedTodos).toContainEqual(newTodo);
    });

    it('Should toggle a todo completion status and update localStorage', async () => {
        const todo = { id: '1', text: 'Todo', isComplete: false };
    
        wrapper.vm.todos = [todo];
        wrapper.vm.currentTodos = [todo];
    
        await wrapper.vm.toggleTodo('1');
    
        expect(wrapper.vm.todos[0].isComplete).toBe(true);
        expect(wrapper.vm.currentTodos[0].isComplete).toBe(true);
    
        const storedTodos = JSON.parse(localStorage.getItem('todos') || '[]');
        expect(storedTodos).toContainEqual({ id: '1', text: 'Todo', isComplete: true });
      });

    it('Should delete a todo and update local storage', async () => {
        const todo = { id: '1', text: 'Todo', isComplete: false };

        wrapper.vm.todos = [todo];
        wrapper.vm.currentTodos = [todo];

        await wrapper.vm.deleteTodo('1');

        const storedTodos = JSON.parse(localStorage.getItem('todos') || '[]');
        expect(storedTodos).toEqual([]);
      });

    it('Should delete completed todos and update local storage', async () => {
        const todos = [
            { id: '1', text: 'Todo 1', isComplete: true },
            { id: '2', text: 'Todo 2', isComplete: false },
            { id: '3', text: 'Todo 3', isComplete: true },
            { id: '4', text: 'Todo 4', isComplete: false },
          ];

        wrapper.vm.todos = todos;
        wrapper.vm.currentTodos = todos;

        await wrapper.vm.deleteCompletedTodos();

        expect(wrapper.vm.todos).toHaveLength(2);
        expect(wrapper.vm.currentTodos).toHaveLength(2);
        expect(wrapper.vm.todos).toEqual([
            { id: '2', text: 'Todo 2', isComplete: false },
            { id: '4', text: 'Todo 4', isComplete: false }
        ]);
        expect(wrapper.vm.currentTodos).toEqual([
            { id: '2', text: 'Todo 2', isComplete: false },
            { id: '4', text: 'Todo 4', isComplete: false }
        ]);

        const storedTodos = JSON.parse(localStorage.getItem('todos') || '[]');
        expect(storedTodos).toHaveLength(2);
        expect(storedTodos).toEqual([
            { id: '2', text: 'Todo 2', isComplete: false },
            { id: '4', text: 'Todo 4', isComplete: false }
        ]);
    });

    it('Should filter current todos based on TodosFilterEnum', async () => {
        const todos = [
            { id: '1', text: 'Todo 1', isComplete: true },
            { id: '2', text: 'Todo 2', isComplete: false }
        ];

        wrapper.vm.todos = todos;
        wrapper.vm.currentTodos = todos;

        await wrapper.vm.changeTodosFilter(TodosFilterEnum.Active);

        expect(wrapper.vm.currentTodos).toHaveLength(1);
        expect(wrapper.vm.currentTodos).toEqual([{ id: '2', text: 'Todo 2', isComplete: false }]);

        expect(wrapper.vm.todos).toHaveLength(2);
        expect(wrapper.vm.todos).toEqual([
            { id: '1', text: 'Todo 1', isComplete: true },
            { id: '2', text: 'Todo 2', isComplete: false }
        ]);
    });

    it('Should update todos order and local storage', async () => {
        const todos = [
            { id: '1', text: 'Todo 1', isComplete: false },
            { id: '2', text: 'Todo 2', isComplete: false }
        ];


        const reorderedTodos = [
            { id: '2', text: 'Todo 2', isComplete: false },
            { id: '1', text: 'Todo 1', isComplete: false }
        ];
        
        wrapper.vm.todos = todos;
        wrapper.vm.currentTodos = todos;

        expect(wrapper.vm.todos).toEqual(todos);
      
        await wrapper.vm.updateTodosOrder(reorderedTodos);

        expect(wrapper.vm.todos).toEqual(reorderedTodos);
      
        const storedTodos = JSON.parse(localStorage.getItem('todos') || '[]');
        expect(storedTodos).toEqual(reorderedTodos);
    });

    it('Should render correct dra and drop text color based on app theme prop', async () => {
        expect(wrapper.vm.dragAndDropTextColor).toBe('#9495A5');
        
        await wrapper.setProps({appTheme: AppThemeEnum.Dark});

        expect(wrapper.vm.dragAndDropTextColor).toBe('#5B5E7E');
    });
})