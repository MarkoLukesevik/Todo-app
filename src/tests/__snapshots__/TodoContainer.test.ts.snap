// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test suite for todo container component > Should match the snapshot 1`] = `
<div
  data-v-app=""
>
  
  <create-todo-stub
    apptheme="0"
    data-v-14f277b8=""
  />
  <todo-list-stub
    apptheme="0"
    data-v-14f277b8=""
    todos=""
    todoscount="0"
    todosfilter="0"
  />
  <todos-action-bar-stub
    apptheme="0"
    data-v-14f277b8=""
    todosfilter="0"
  />
  <p
    class="drag-and-drop"
    data-v-14f277b8=""
    style="color: rgb(148, 149, 165);"
  >
     Drag and drop to reorder list 
  </p>
  
</div>
`;
